### ISSUE TITLE
reuse of a prng key in expand_once()

### ISSUE THREAD

**Issue by wiep** [blackjax-devs/blackjax#53]
in the function expand_once a PRNG key is used multiple times.
https://github.com/blackjax-devs/blackjax/blob/6938da9e7ce4aa952e8c0b44666616c46acf3b68/blackjax/inference/trajectory.py#L507

in particular, the key `rng_key` generated in line 523 is used twice: in the call to trajectory_integrator (line 540) and as an element of the loop state returned in line 588. thus, it is reused in the next iteration for splitting. as far as i understand jax' rng implementation this kind of reuse must be avoided.

**Comment by rlouf** [comment]
I think that it is fine as long as the same key is not used at two distinct places. I don't think it's the case because the key that's passed in the while loop is then split?

**Comment by wiep** [comment]
my understanding is that splitting should 'consume' the key. however, I am not certain.

if i understood you argument correctly, this code should be correct:

```py
def f(rng_key):
    u = jax.random.normal(rng_key)
    rng_key, _ = jax.random.split(rng_key, 2) # reuse of key is ok, when used for splitting.
    v =  jax.random.normal(rng_key)
    return(u, v)

key = jax.random.PRNGKey(0)
print(f(key))
key, _ = jax.random.split(key, 2) # reuse of key is ok, when used for splitting.
print(f(key))
```
it generates the same numbers in for `v` in the first call and `u` in the second call, though.

**Comment by rlouf** [comment]
That's the expected behavior in the example you provide. Now, if we strip away the unnecessary code in trajectory extension we have:

```python
def loop(rng_key, state):
    rng_key_new, direction_key, proposal_key = jax.random.split(rng_key, 3)
    direction = jnp.where(jax.random.bernoulli(direction_key), 1, -1)  
    state = trajectory_integrator(rng_key_new, direction, state)
    return rng_key_new, state
```

So `rng_key_new` is never re-used since it gets split immediately at the next call of `loop`.

**Comment by wiep** [comment]
i think the issue exists if one cannot rule out the possibility that `rng_key` is split twice. consider this example that might be close to whats going on in expand_once().
```py
def f(rng_key):
    rng_key, snd_key = jax.random.split(rng_key)
    u = jax.random.uniform(snd_key)
    v, w = g(rng_key) # should mimic the call of trajectory_integrator
    return rng_key, (u, v, w)

def g(rng_key):
    keys = jax.random.split(rng_key)
    v = jax.random.uniform(keys[0])
    w = jax.random.uniform(keys[1])
    return (v, w)

rng_key = jax.random.PRNGKey(0)
rng_key, tupl0 = f(rng_key)
_, tupl1 = f(rng_key)

print(tupl0[2],  tupl1[0])
```

`rng_key` is split in the beginning of each function. the same value for the rng key is used in the first call to `g` and in the second call to `f`. therefore, the last entry in the tuple returned by the first call of `f` equals the first entry in the tuple returned in the second call of `f`.

couldn't something similar happen in the body of `trajectory_integrator`; e.g. using `split(rng_key, 3)`?

**Comment by rlouf** [comment]
You're right, we should make sure inside the function that there is no way this could happen.

**Comment by rlouf** [comment]
I saw you mentioned the issue in a commit on your fork, would you like to submit a PR?

**Comment by wiep** [comment]
i thought i already opened a PR; apparently, i didn't... it should be opened now.

**Comment by rlouf** [comment]
Fixed by #55
